File: casbot2_cpp_demo/src/basic_control_demo.cpp
#include <chrono>
#include <memory>
#include <string>
#include <vector>

#include "geometry_msgs/msg/twist.hpp"
#include "rclcpp/rclcpp.hpp"
#include "std_srvs/srv/set_bool.hpp"

using namespace std::chrono_literals;

class BasicControlDemo : public rclcpp::Node
{
public:
  BasicControlDemo() : Node("casbot2_basic_control_demo")
  {
    nav_mode_cli_ = create_client<std_srvs::srv::SetBool>("/motion/switch_nav_mode");
    cmd_vel_pub_ = create_publisher<geometry_msgs::msg::Twist>("/navigation/cmd_vel", 10);
  }

  bool switch_nav_mode(bool enable_nav)
  {
    if (!nav_mode_cli_->wait_for_service(3s)) {
      RCLCPP_ERROR(get_logger(), "Service /motion/switch_nav_mode not available");
      return false;
    }
    auto req = std::make_shared<std_srvs::srv::SetBool::Request>();
    req->data = enable_nav;
    auto future = nav_mode_cli_->async_send_request(req);
    if (rclcpp::spin_until_future_complete(shared_from_this(), future, 3s) !=
      rclcpp::FutureReturnCode::SUCCESS) {
      RCLCPP_ERROR(get_logger(), "Failed to call /motion/switch_nav_mode");
      return false;
    }
    const auto resp = future.get();
    RCLCPP_INFO(get_logger(), "switch_nav_mode(%s): success=%s, msg=%s",
      enable_nav ? "NAVIGATION" : "IDLE_MODE",
      resp->success ? "true" : "false",
      resp->message.c_str());
    return resp->success;
  }

  void send_cmd_vel(double vx, double wz)
  {
    geometry_msgs::msg::Twist msg;
    msg.linear.x = vx;
    msg.angular.z = wz;
    cmd_vel_pub_->publish(msg);
  }

private:
  rclcpp::Client<std_srvs::srv::SetBool>::SharedPtr nav_mode_cli_;
  rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr cmd_vel_pub_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<BasicControlDemo>();

  if (!node->switch_nav_mode(true)) {
    rclcpp::shutdown();
    return 1;
  }
  rclcpp::Rate rate(10.0);
  for (int i = 0; i < 20 && rclcpp::ok(); ++i) {
    node->send_cmd_vel(0.2, 0.0);
    rclcpp::spin_some(node);
    rate.sleep();
  }
  for (int i = 0; i < 10 && rclcpp::ok(); ++i) {
    node->send_cmd_vel(0.0, 0.0);
    rclcpp::spin_some(node);
    rate.sleep();
  }
  const bool exited = rclcpp::ok() && node->switch_nav_mode(false);

  rclcpp::shutdown();
  return exited ? 0 : 1;
}

File: casbot2_py_demo/casbot2_py_demo/basic_control_demo.py
import time

import rclpy
from geometry_msgs.msg import Twist
from rclpy.node import Node
from std_srvs.srv import SetBool



class BasicControlDemo(Node):
    def __init__(self) -> None:
        super().__init__("casbot2_py_control_demo")
        self.nav_mode_cli = self.create_client(SetBool, "/motion/switch_nav_mode")
        self.cmd_vel_pub = self.create_publisher(Twist, "/navigation/cmd_vel", 10)

    def switch_nav_mode(self, enable_nav: bool) -> bool:
        if not self.nav_mode_cli.wait_for_service(timeout_sec=3.0):
            self.get_logger().error("Service /motion/switch_nav_mode not available")
            return False
        req = SetBool.Request()
        req.data = enable_nav
        future = self.nav_mode_cli.call_async(req)
        rclpy.spin_until_future_complete(self, future, timeout_sec=3.0)
        resp = future.result()
        if resp is None:
            self.get_logger().error("Failed to call /motion/switch_nav_mode")
            return False
        target = "NAVIGATION" if enable_nav else "IDLE_MODE"
        self.get_logger().info(
            f"switch_nav_mode({target}): success={resp.success}, msg={resp.message}"
        )
        return bool(resp.success)

    def send_cmd_vel(self, vx: float, wz: float) -> None:
        msg = Twist()
        msg.linear.x = float(vx)
        msg.angular.z = float(wz)
        self.cmd_vel_pub.publish(msg)


def main() -> None:
    rclpy.init()
    node = BasicControlDemo()

    entered = False
    exited = False
    try:
        entered = node.switch_nav_mode(True)
        if not entered:
            return 1
        for _ in range(20):
            if not rclpy.ok():
                break
            node.send_cmd_vel(0.2, 0.0)
            rclpy.spin_once(node, timeout_sec=0.0)
            time.sleep(0.1)
    finally:
        if entered and rclpy.ok():
            for _ in range(10):
                node.send_cmd_vel(0.0, 0.0)
                rclpy.spin_once(node, timeout_sec=0.05)
            exited = node.switch_nav_mode(False)
            if not exited:
                node.get_logger().error('退出导航失败，请检查 /motion/status')
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()
    return 0 if exited else 1


if __name__ == "__main__":
    raise SystemExit(main())
